src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,355 @@
+        let mut params: HashMap<&str, String> = HashMap::new();
+        params.insert("grant_type", REFRESH_TOKEN_GRANT_TYPE.to_string());
+        params.insert("refresh_token", req.refresh_token);
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// performs an oauth2 token exchange with the provided [ExchangeTokenRequest] information.
+    async fn exchange_token(&self, req: ExchangeTokenRequest) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+
+        params.insert("grant_type", TOKEN_EXCHANGE_GRANT_TYPE.to_string());
+        params.insert("requested_token_type", ACCESS_TOKEN_TYPE.to_string());
+
+        params.insert("subject_token", req.subject_token);
+        params.insert("subject_token_type", req.subject_token_type);
+
+        params.insert("scope", req.scope.join(" "));
+
+        if let Some(audience) = req.audience {
+            params.insert("audience", audience);
+        }
+        if let Some(resource) = req.resource {
+            params.insert("resource", resource);
+        }
+        if let Some(actor_token) = req.actor_token {
+            params.insert("actor_token", actor_token);
+        }
+        if let Some(actor_token_type) = req.actor_token_type {
+            params.insert("actor_token_type", actor_token_type);
+        }
+
+        //TODO(aviebrantz): Add support for extra options
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// execute http request and token exchange
+    pub(crate) async fn execute(
+        &self,
+        url: String,
+        client_auth: ClientAuthentication,
+        headers: http::HeaderMap,
+        params: HashMap<&str, String>,
+    ) -> Result<TokenResponse> {
+        let mut headers = headers.clone();
+        let mut params = params.clone();
+        println!("[execute] url: {}", url);
+        client_auth.inject_auth(&mut headers, &mut params);
+        println!("[execute] headers: {:?}", headers);
+
+        let res = self
+            .client
+            .post(url)
+            .form(&params)
+            .headers(headers)
+            .send()
+            .await
+            .map_err(Error::io)?;
+
+        println!("[execute] response: {:?}", res);
+        let token_res = res.json::<TokenResponse>().await.map_err(Error::io)?;
+        Ok(token_res)
+    }
+}
+
+/// TokenResponse is used to decode the remote server response during
+/// an oauth2 token exchange.
+#[derive(Deserialize, Default, PartialEq, Debug)]
+pub(crate) struct TokenResponse {
+    access_token: String,
+    issued_token_type: String,
+    token_type: String,
+    expires_in: u64,
+    scope: String,
+    refresh_token: Option<String>,
+}
+
+/// Authentication style via headers or form params.
+#[derive(Debug, Clone)]
+enum ClientAuthStyle {
+    Unknown,
+    InParams,
+    InHeader,
```

Comment:
In the spec, I see the Client Auth in header. Does spec mention somewhere about sending the Client Auth in parameters?

Reply:
here the spec mentions that the authentication can go on the header with the basic auth format or in the request body with `client_id`/`client_secret`

https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1

Reply:
I see. Are we going to use it in params though? If our STS supports headers, we would use that always?

Reply:
The Java and Go SDK use in params for 3 Legged Oauth, I'm trying to find the spec for that. In any case, as mentioned on the RFC, it is possible to go in the request body, so I think it's ok to support here.

Reply:
Could you please point me to where in params is used in Java?

I see auth is sent in header for [ExternalAccountAuthorizedUser](https://github.com/sai-sunder-s/google-auth-library-java/blob/main/oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java#L365-L371). I am assuming it is the same for all cred types?

Reply:
Sorry, I found such usage in Python and Go, not Java. Which was corroborated by the aforementioned RFC https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1

In the Go auth they are used in the 3 Legged Oauth flow like I mentioned:
* https://github.com/googleapis/google-cloud-go/blob/main/auth/credentials/filetypes.go#L159
* https://github.com/googleapis/google-cloud-go/blob/main/auth/threelegged.go#L291

For Python:
* https://github.com/googleapis/google-auth-library-python/blob/56b4d3b3bfd6f935411130c8828cdeea90aa69f1/google/oauth2/utils.py#L93

Reply:
The go code you referred to is for User Credentials (We need to send in params for token endpoint call), it is not for byoid.

Python does seem to use it for byoid but I am not sure if that case ever hits. We have learnt that python code usually has lots of unnecessary stuff.

I think it is better that we support only auth through headers if that is accepted by STS. @lsirac to confirm that sending auth through header is sufficient for all the current byoid flows.